test/common/tcp_proxy/tcp_proxy_test.cc

```diff
@@ -326,18 +336,86 @@ TEST(ConfigTest, DEPRECATED_FEATURE_TEST(Routes)) {
     connection.remote_address_ =
         std::make_shared<Network::Address::Ipv6Instance>("2003:0:0:0:0::5");
     EXPECT_EQ(std::string("with_v6_source_and_destination"),
-              config_obj.getRouteFromEntries(connection));
+              config_obj.getRouteFromEntries(connection)->clusterName());
   }
 
   {
     // fall through
     NiceMock<Network::MockConnection> connection;
     connection.local_address_ = std::make_shared<Network::Address::Ipv6Instance>("2004::");
     connection.remote_address_ = std::make_shared<Network::Address::Ipv6Instance>("::");
-    EXPECT_EQ(std::string("catch_all"), config_obj.getRouteFromEntries(connection));
+    EXPECT_EQ(std::string("catch_all"), config_obj.getRouteFromEntries(connection)->clusterName());
   }
 }
 
+TEST(ConfigTest, WeightedClusterWithZeroWeightConfig) {
+  const std::string yaml = R"EOF(
+  stat_prefix: name
+  weighted_clusters:
+    clusters:
+    - name: cluster1
+      weight: 1
+    - name: cluster2
+)EOF";
+
+  NiceMock<Server::Configuration::MockFactoryContext> factory_context;
+  EXPECT_THROW(constructConfigFromV2Yaml(yaml, factory_context), EnvoyException);
+}
+
+TEST(ConfigTest, WeightedClustersConfig) {
+  const std::string yaml = R"EOF(
+  stat_prefix: name
+  weighted_clusters:
+    clusters:
+    - name: cluster1
+      weight: 1
+    - name: cluster2
+      weight: 2
+)EOF";
+
+  NiceMock<Server::Configuration::MockFactoryContext> factory_context;
+  Config config_obj(constructConfigFromV2Yaml(yaml, factory_context));
+
+  NiceMock<Network::MockConnection> connection;
+  EXPECT_CALL(factory_context.random_, random()).WillOnce(Return(0));
+  EXPECT_EQ(std::string("cluster1"), config_obj.getRouteFromEntries(connection)->clusterName());
+
+  EXPECT_CALL(factory_context.random_, random()).WillOnce(Return(2));
+  EXPECT_EQ(std::string("cluster2"), config_obj.getRouteFromEntries(connection)->clusterName());
+}
+
+TEST(ConfigTest, RouteMetadataMatchConfig) {
```

Comment:
I think this is the basically the same as `WithMetadataMatch`. I think we need one that defines metadata match on weighted cluster entries.

Reply:
done